### DISCUSSION TITLE
Feature request: Intersection Polyline

### DISCUSSION THREAD

**Discussion by yetigit** [elalish/manifold#116] [Q&A]
Hi, thank you for your efforts on doing a fast and robust csg lib 

what would be a killer feature is being able to retrieve the intersection data, precisely as naturally ordered polyline(s). 
which boils down to just having the intersection points in the right order ; with a topological predicate since there can be several intersection loops.

here a sphere vs a cube
![maya_zk2js5vmOH](https://user-images.githubusercontent.com/7016200/157231121-0569c059-6335-4fd1-97c9-b23ec30e7b7b.png)

cube vs cube
![maya_qHK3Isv83y](https://user-images.githubusercontent.com/7016200/157231147-5d130856-2332-4dc1-bdfc-65049ad7b303.png)

**Comment by elalish** [comment]
Check out my [GetMeshRelation()](https://elalish.github.io/manifold/classmanifold_1_1_manifold.html#a45ea90568e3c9072089e9854cd51a02f) function. It returns this data (in a different format, suitable for multiple uses like texture mapping), even after a whole series of Boolean operations. It should be pretty simple to convert it to the polyline format you desire; feel free to add a PR!

**Reply by dchasman** [reply]
Hi @elalish - I had the identical question but it appears the answer above is outdated (link referencing _GetMeshRelation()_ is 404’ing)? Is there a modern Manifold API way to accomplish this?

**Reply by elalish** [reply]
Yes, we have a better way: [`GetMeshGL`](https://manifoldcad.org/docs/html/structmanifold_1_1_mesh_g_l.html). This has a more complete representation of the relationships.

**Reply by elalish** [reply]
This is the only kind of `mesh` in our JS and Python bindings.

**Reply by hrgdavor** [reply]
I am still on the sidelines watching manifold progress, and have used it only a bit, but I am wondering about a thing regarding union and filet.

There is a use case when unioning two shapes to filet the intersection edge like this issue: https://github.com/openscad/openscad/issues/884  ...

Is it possible to get something in that direction by finding the edge after union and using smooth in manifold ? 
if possible, it would make for a great example script on monifoldcad web.

**Reply by elalish** [reply]
Yes, part of why I developed our smoothing algorithms is to eventually enable features like fillet. However, it still needs considerable work: beyond just identifying the intersection edges, you'll also need to create new vertices that represent the edges of the fillet and remove the actual intersection verts. Basically we need a `MakeChamfer` function first, then from that we can use smoothing to turn the chamfer into a fillet.

**Reply by hrgdavor** [reply]
@elalish that is great to hear. I am glad you are looking to tackling this eventually :)

**Reply by elalish** [reply]
If you know anything about chamfer algorithms, I'm all ears. I think that will still be pretty hard.

**Reply by hrgdavor** [reply]
aah, sorry :) my knowledge is lacking in computational geometry. Best I can do for now is to cheer you :) and hope you can find something workable for chamfer.

**Comment by ofmetal** [comment]
So, what is the best way to get that intersection polyline in late 2024 ?

**Reply by elalish** [reply]
`GetMeshGL()`, pair up the halfedges (I wonder if it would be enough added convenience to expose these someday?), find any pairs whose triangles are from different `runs` - these are your intersection edges. Connect the halfedges from one run head-to-tail to form the polylines.

**Reply by ofmetal** [reply]
The halfedges - the ones that could be extracted from **triVerts** array?
Plus triangle's parent mesh id from the **runIndex** array? Or is it **runOriginalID**?

**Reply by elalish** [reply]
Yes and yes. The `runOriginalID` tells you the parent mesh id for each run. If you happen to intersect two copies of the same mesh, that ID will be the same, so safest to just check which part of the `runIndex` array you're in - as those will always be distinct even if they correspond to the same ID.
